# Feedback from previous weeks and other and hints

1. Be careful about your environment remembering variables. Make sure your code works in a new _clean_ environment. In Colab: `Runtime`->`restart Runtime`, in Anaconda's Jupyter: `Kernel`->`Restart`.
2. Graphs without labels (or units when appropriate) are not worth any point.
3. Do put in sufficient explanatory comments in your code.
4. Functions are very important. Do look up the video on the Safari O'Reilly ressource if you are still not clear on them !

For this week you can use these imports at the start of your programs:

In [36]:
import numpy as np
import matplotlib.pyplot as plt

We will use a new module `uncertainties`, which is **not** standard in the colab environment. You will have to first run:

In [10]:
! pip install -q uncertainties

To install the module before you can import it:

In [38]:
import uncertainties as uc
import uncertainties.umath as um # for maths functions

# Introduction
In the practical classes PX2133/PX2233 and PX2338 (Obs tech), as well as your year 3/4 project, a lot of emphasis is placed on the determination and mathematical handling of errors.
The uncertainties module allows us to deal very easily with [error propagation](https://en.wikipedia.org/wiki/Propagation_of_uncertainty). For this sheet you should remind yourself about error bars in measurements and about propagation of uncertainties. Take an example from your lab handbook:

**Example 1**: If the length of a rectangle is $1.24\pm0.02 m$ and its breadth is $0.61\pm0.01 m$, what is its area and the error in the area? The following code snippet solves this problem in a few lines.

In [ ]:
L = uc.ufloat(1.24, 0.02)
W = uc.ufloat(0.61, 0.01)
print ('Area is:', L*W, 'm^2') # Do remember to add the units when printing!

Area is: 0.756+/-0.017 m^2


**Note**: For the area itself, it's fairly straightforward:

In [ ]:
1.24*0.61

0.7564

However, for the error bar on this number:

In [ ]:
0.02*0.01

0.0002

does not work. Instead, the [error progation formula](https://en.wikipedia.org/wiki/Propagation_of_uncertainty#Example_formulae) gives:

In [ ]:
np.abs(1.24*0.61)*np.sqrt((0.02/1.24)**2+(0.01/0.61)**2)

0.01739540169125163

So the area is $0.756\pm0.017 m^2$. `uncertainties` obviously saves a lot of work, even for such a simple case. You can also take a look at the web site uncertainties hosted at https://pythonhosted.org/uncertainties/user_guide.html. In particular, [this section](https://pythonhosted.org/uncertainties/user_guide.html#access-to-the-uncertainty-and-to-the-nominal-value) shows some of the properties of a `ufloat` you can access directly.

**Example 2**: A reference object is $10.0\pm0.0001 m$ long, and makes a viewing angle of $0.62\pm0.02 rad$. How far is it?

In [ ]:
L = uc.ufloat(10.0, 0.0001)
theta = uc.ufloat(0.62,0.02)

Distance = (L/2)/um.tan(theta/2)

print ('Distance is:', Distance.nominal_value, 'm, with an error of:', Distance.std_dev)

Distance is: 15.609024890896208 m, with an error of: 0.537283338762715


Note the need to use "umath" functions (like `um.tan()` instead of `np.tan()`), and how to get the nominal value and the standard deviation of the uncertainties objects. To get nicer looking output, such as controlling the number of significant digits printed, you can use the information about formatting at https://docs.python.org/3/tutorial/inputoutput.html. In the exercises below you need to print the values to the screen. (Don’t forget units.)

# Exercises
This must be marked before you leave the lab. Mark weighting is in brackets.
**Save your work to GitHub after having run all cells with `Runtime` -> `Restart and run all`. And do not change the notebook's filename.** Do add comments to your code, you'll lose points if your code is hard to understand. Graphs without labels (or units when appropriate) are not worth any point.

## Exercise 0
[0] With some approximations, we have measured the mass of the following black-holes:
```
"35.6+/-3.9","30.6+/-3.7","63.1+/-3.2","23.2+/-9.8","13.6+/-4.5","35.7+/-6.8","13.7+/-6.0","7.7+/-2.4","20.5+/-4.0"
```
Compute for each (with error-bars) their lifetime due to Hawking radiation:
$$
t = \left(\frac{M}{M_{\odot}}\right)^3\,\times\,2.097\,\times\,10^{67} yr
$$

(this exercise is for demonstration purposes and won't be marked)

In [2]:
import numpy as np
import matplotlib.pyplot as plt
! pip install -q uncertainties
import uncertainties as uc
import uncertainties.umath as um # for maths functions



   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.1/60.1 kB 1.9 MB/s eta 0:00:00


In [4]:
#Define the solar mass in the same units as the black hole masses (assuming solar masses)
Msolar = uc.ufloat(1.0, 0.0) #Assuming solar mass has negligible uncertainty for this calculation

#Black hole masses with uncertainties as strings
blackholemassess = ["35.6+/-3.9","30.6+/-3.7","63.1+/-3.2","23.2+/-9.8","13.6+/-4.5","35.7+/-6.8","13.7+/-6.0","7.7+/-2.4","20.5+/-4.0"]

#Constant for the lifetime calculation
lifetimeconstant = 2.097e67 # yr

#Convert the string representations of masses to ufloat objects
blackholemassesusfloat = []
for massstr in blackholemassess:
    value, uncertainty = massstr.split('+/-')
    blackholemassesusfloat.append(uc.ufloat(float(value), float(uncertainty)))


#Calculate the lifetime for each black hole
blackholelifetimes = [(mass / Msolar)**3 * lifetimeconstant for mass in blackholemassesusfloat]

#Print the results
print("Black Hole Lifetimes due to Hawking Radiation:")
for i, lifetime in enumerate(blackholelifetimes):
    print(f"Black Hole {i+1}: {lifetime:.3e} yr") #Formatting to scientific notation with 3 decimal places and units

Black Hole Lifetimes due to Hawking Radiation:
Black Hole 1: (9.461+/-3.109)e+71 yr
Black Hole 2: (6.008+/-2.180)e+71 yr
Black Hole 3: (5.268+/-0.802)e+72 yr
Black Hole 4: (2.619+/-3.318)e+71 yr
Black Hole 5: (5.275+/-5.236)e+70 yr
Black Hole 6: (9.541+/-5.452)e+71 yr
Black Hole 7: (5.392+/-7.085)e+70 yr
Black Hole 8: (9.573+/-8.952)e+69 yr
Black Hole 9: (1.807+/-1.058)e+71 yr


/usr/local/lib/python3.12/dist-packages/uncertainties/core.py:1024: UserWarning: Using UFloat objects with std_dev==0 may give unexpected results.
  warn("Using UFloat objects with std_dev==0 may give unexpected results.")


## Exercise 1
[2] An object is measured to travel a distance $x = 5.1 \pm 0.4 m$ during a time of $t = 0.4 \pm 0.1 s$. What is the average velocity and the error in the average velocity?

In [4]:
#Define distance and time
import uncertainties as uc
x = uc.ufloat(5.1,0.4) 
t = uc.ufloat(0.4,0.1)
v = x/t
print(v)


12.7+/-3.3


## Exercise 2
[2] An enterprising cow attempts to jump over the moon by jumping vertically into the air with initial speed $v_0=4.0\pm0.2 m/s$. After a time $t=0.60\pm0.06s$, the height of the cow is $h = v_0t-\frac{1}{2}g t^2 = 0.636 m$. What is the uncertainty in $h$? Take $g$ as exactly $9.81 ms^{-2}$.

In [8]:
#Define initial velocity and time
import uncertainties as uc
v0 = uc.ufloat(4,0.2)
t = uc.ufloat(0.6,0.06)
h = v0*t - 0.5*9.81*t**2
print(h)
print('uncetainty is +/- 0.16')

0.63+/-0.16
uncetainty is +/- 0.16


## Exercise 3
[2] In an optics experiment the object distance $u$ is measured to be 20cm and the image distance $v$ is 10cm, both to an accuracy of 0.5cm. Find the focal length $f$ of the lens using the formula:

$$ \frac{1}{u}+\frac{1}{v}=\frac{1}{f}$$

In [12]:
#Define object distance and image distance
import uncertainties as uc
u = uc.ufloat(20,0.5)
v = uc.ufloat(10,0.5)
f = (1/u + 1/v)**-1
print(f)


6.67+/-0.23


## Exercise 4
[2] Two students each measure the refractive index of water. Jack measures a value of $1.33 \pm 0.03$ while Jill measures $1.28 \pm 0.02$. Are these values in agreement? *You do have to think a bit about this one...*

In [22]:
#Define refractive index measurements with uncertainties
jack = uc.ufloat(1.33,0.03)
jill = uc.ufloat(1.28,0.02)
diff = jack - jill
print(diff)
print("difference is larger than the two uncertainties so not in agreement")

0.05+/-0.04
difference is larger than the two uncertainties so not in agreement


## Exercise 5
[2] The damped resonance frequency $\omega_{res}$ of an oscillating system is related to the (un-damped) natural angular frequency $\omega_0$ and the damping coefficient $\alpha$ by:

$$\omega_{res} = \sqrt{ \omega_0^2 - 2\alpha^2}$$

Find $f_0$ if the measured resonance frequency $f_{res}$ is $23.2\pm0.1 Hz$ and the measured damping coefficient is $19.5\pm0.5s^{-1}$.


In [40]:
#Define measured resonance frequency with uncertainty in Hz
wres = uc.ufloat(23.2,0.1)
damp = uc.ufloat(19.5,0.5)
f0 = um.sqrt(wres**2+2*damp**2)
print(f0)

36.0+/-0.5


## Exercise 6
[2] Suppose you have the following equation from one of your lab experiments:

$$f=\frac{c}{2}\sqrt{\frac{n_x^2}{L_x^2}+\frac{n_y^2}{L_y^2}+\frac{n_z^2}{L_z^2}}$$

where $f$ is the resonant frequency of sound waves in a box of sides $L_x$, $L_y$ and $L_z$ in length and the $n_x$ etc. are integers. $L_x = 10.2\pm0.2m$, $L_y = 5.2\pm0.3m$ and $L_z = 20.0\pm0.1 m$, while $c = 331.3 + T * 0.606 \,m\,s^{-1}$ is the temperature-dependent speed of sound, and the temperature $T$ is $23 \pm 1^\circ C$.
Calculate $f$ and the error in $f$ for the following values of $(nx,ny,nz)= (1,1,1), (1,1,2)$ and $(2,1,1)$.


In [42]:
#Define dimensions of box with uncertainties
Lx = uc.ufloat(10.2, 0.2)
Ly = uc.ufloat(5.2,0.3)
Lz = uc.ufloat(20,0.1)
T = uc.ufloat(23, 1)
c = (331.3+T*0.606)

def f(Nx,Ny,Nz):
    f = c/2 *um.sqrt( (Nx**2/Lx**2) + (Ny**2/Ly**2) + (Nz**2/Lz**2))
    return f

f111 = f(1,1,1)
f112 = f(1,1,2)
f211 = f(2,1,1)

print(f111 , f112, f211)

38.2+/-1.7 41.1+/-1.6 48.2+/-1.4


## Exercise 7
[4] The reflection coefficient $R_\parallel$ for parallel plane-polarised light reflected from a surface is given by the equation:

$$ R_\parallel = \frac{\tan^2(\theta_i - \theta_t)}{\tan^2(\theta_i + \theta_t)} $$

Calculate the error in $R_\parallel$ given measurements $\theta_i = (78 \pm 1)^\circ$ and $\theta_t = (40 \pm 1)^\circ$.

In [62]:
import uncertainties.umath as um
oi = uc.ufloat(78,1)
ot = uc.ufloat(40,1)

def Rp(ideg, ideg):
    ti = ideg * np.pi/180
    tt = tdeg *np.pi/180
    
    R = (um.tan(ti - tt))**2/(um.tan(ti+tt))**2
    return (R)

h = Rp(oi, ot)
print(h)

0.173+/-0.027


In [16]:
# Define the parameters with uncertainties and convert units where necessary
#mass flow rate
mdot = uc.ufloat(0.13, 0.01) # kg/s
#orifice diameter
d = uc.ufloat(11, 1) # mm
D1 = d/1000 # m
#pipe diameter
D = uc.ufloat(71, 1) # mm
D2 = D/1000 # m
#fluid density
rho_gcm3 = uc.ufloat(1.01, 0.01) # g/cm^3
rho = rho_gcm3 * 1000 # kg/m^3 (1 g/cm^3 = 1000 kg/m^3)
#differential pressure
deltaP = uc.ufloat(156, 7) # Pa
#K and F are constants with no units
K = uc.ufloat(1.0, 0.0)
F = uc.ufloat(1.0, 0.0)

#Calculate discharge coefficient (Cd) using  given formula
#Cd = (mdot * um.sqrt(1 - (d/D)**4)) / (K * d**2 * F * um.sqrt(rho * deltaP))
#Since K and F are constants with no units can simplify the denominator
Cd = (mdot * um.sqrt(1 - (D1/D2)**4)) / (D1**2 * um.sqrt(rho * deltaP))


#Calculate fractional uncertainty as a percentage
#Fractional uncertainty = (uncertainty in Cd / nominal value of Cd) * 100
fractional_uncertainty_Cd = (Cd.std_dev / abs(Cd.nominal_value)) * 100

#Print fractional uncertainty formatted as percentage to one significant figure
print(f"The fractional uncertainty in the discharge coefficient (Cd) is: {fractional_uncertainty_Cd:.1g}%")

The fractional uncertainty in the discharge coefficient (Cd) is: 2e+01%


/usr/local/lib/python3.12/dist-packages/uncertainties/core.py:1024: UserWarning: Using UFloat objects with std_dev==0 may give unexpected results.
  warn("Using UFloat objects with std_dev==0 may give unexpected results.")


## Exercise 9: Optional problem (not marked)
If you have time and want to try something interesting, do the following problem by plotting in 2D:
 - Draw an equilateral triangle with vertices and coordinates: vertex 1: $(p_1,q_1)$; vertex 2: $(p_2, q_2)$; vertex 3: $(p_3, q_3)$.
 - Place a dot at an arbitrary point $P = (x_0, y_0)$ within this triangle.
 - Find the next point by selecting randomly an integer $n = 1 , 2, $  or $3$ :
    1. If 1 , place a dot halfway between P and vertex 1.
    2. If 2 , place a dot halfway between P and vertex 2.
    3. If 3 , place a dot halfway between P and vertex 3.
 - Repeat the last two steps using the last dot as the new P.

Mathematically, the coordinates of successive points are given by the formulae

$$(x_{i+1},y_{i+1})=0.5[(x_i,y_i)+(p_n,q_n)]$$

and

$$n=int(1+3r_i),$$

where $r_i$ is a random number between 0 and 1 and where the $int()$ function outputs the closest integer smaller than or equal to the argument.

Try extending this to four vertices.

In [ ]:
mport numpy as np
import matplotlib.pyplot as plt

# 0. LOADING DATA FILES (VERY COMMON IN EXAMS)
# ALWAYS CHECK:
# - File format (.txt, .csv)
# - Column order
# - Whether there is a header line
# - What the columns physically represent

data = np.loadtxt("data.txt")      # or "data.csv"
# data.shape → (N, num_columns)

# Access columns
x = data[:, 0]
y = data[:, 1]

# ---- File WITH header line ----
data = np.loadtxt("data.txt", skiprows=1)

# ---- CSV with named columns (VERY USEFUL) ----
# If allowed to use pandas:
#
# import pandas as pd
# df = pd.read_csv("data.csv")
# x = df["time"]
# y = df["velocity"]
# yerr = df["uncertainty"]

#CREATING NUMBERS WITH UNCERTAINTY
a = ufloat(2.0, 0.1)   # 2.0 ± 0.1
b = ufloat(3.0, 0.2)   # 3.0 ± 0.2

#USING umath 
d1 = um.sqrt(a)
d2 = um.sin(a)
d3 = um.exp(a)
d4 = um.log(b)


#ARRAYS OF MEASUREMENTS WITH UNCERTAINTY

# Create ufloats element-by-element (no unumpy used)
y_u = [ufloat(val, err) for val, err in zip(y, yerr)]

# Example transformation
z_u = [2.0 * yi + 1.0 for yi in y_u]
# Extract nominal values and uncertainties
z = np.array([zi.nominal_value for zi in z_u])
zerr = np.array([zi.std_dev for zi in z_u])

#COMMON PHYSICS EXAMPLE
# Density: ρ = m / V

m = ufloat(0.250, 0.005)   # kg
V = ufloat(1.2e-4, 0.1e-4) # m^3

rho = m / V
# rho has automatically propagated uncertainty

#CURVE FITTING + UNCERTAINTIES
# scipy.optimize.curve_fit returns:
# popt → best-fit parameters
# pcov → covariance matrix

def model(x, A, B):
    return A * x + B

popt, pcov = curve_fit(model, x, y, sigma=yerr, absolute_sigma=True)

A_fit, B_fit = popt
A_err, B_err = np.sqrt(np.diag(pcov))

# Convert fit results into uncertainty objects
A_u = ufloat(A_fit, A_err)
B_u = ufloat(B_fit, B_err)

y_pred_u = A_u * x + B_u







